src/Compilers/CSharp/Test/Semantic/Semantics/RawInterpolationTests.cs

```diff
@@ -0,0 +1,1276 @@
+    }
+}";
+        CreateCompilationWithMscorlib40AndSystemCore(text).VerifyDiagnostics();
+    }
+
+    [Fact, WorkItem(1098612, "http://vstfdevdiv:8080/DevDiv2/DevDiv/_workitems/edit/1098612")]
+    public void MissingConversionFromFormattableStringToIFormattable()
+    {
+        var text =
+@"namespace System.Runtime.CompilerServices
+{
+    public static class FormattableStringFactory
+    {
+        public static FormattableString Create(string format, params object[] arguments)
+        {
+            return null;
+        }
+    }
+}
+
+namespace System
+{
+    public abstract class FormattableString
+    {
+    }
+}
+
+static class C
+{
+    static void Main()
+    {
+        System.IFormattable i = $""""""{""""}"""""";
+    }
+}";
+        CreateCompilationWithMscorlib40AndSystemCore(text).VerifyEmitDiagnostics(
+                // (23,33): error CS0029: Cannot implicitly convert type 'FormattableString' to 'IFormattable'
+                //         System.IFormattable i = $"""{""}""";
+                Diagnostic(ErrorCode.ERR_NoImplicitConv, @"$""""""{""""}""""""").WithArguments("System.FormattableString", "System.IFormattable").WithLocation(23, 33));
+    }
+
+    [Fact]
+    public void InterpolatedStringBeforeCSharp6()
+    {
+        var text = @"
+class C
+{
+    string M()
+    {
+        return $""""""hello"""""";
+    }
+}";
+
+        CreateCompilation(text, parseOptions: CSharpParseOptions.Default.WithLanguageVersion(LanguageVersion.CSharp5)).VerifyDiagnostics(
+            // (6,16): error CS8026: Feature 'interpolated strings' is not available in C# 5. Please use language version 6 or greater.
+            //         return $"""hello""";
+            Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion5, @"$""""""hello""""""").WithArguments("interpolated strings", "6").WithLocation(6, 16));
+    }
+
+    [Fact]
+    public void InterpolatedStringWithReplacementBeforeCSharp6()
+    {
+        var text = @"
+class C
+{
+    string M()
+    {
+        string other = """"""world"""""";
+        return $""""""hello + {other}"""""";
+    }
+}";
+
+
+        CreateCompilation(text, parseOptions: CSharpParseOptions.Default.WithLanguageVersion(LanguageVersion.CSharp5)).VerifyDiagnostics(
+            // (6,24): error CS8652: The feature 'raw string literals' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+            //         string other = """world""";
+            Diagnostic(ErrorCode.ERR_FeatureInPreview, @"""""""world""""""").WithArguments("raw string literals").WithLocation(6, 24),
+            // (7,16): error CS8026: Feature 'interpolated strings' is not available in C# 5. Please use language version 6 or greater.
+            //         return $"""hello + {other}""";
+            Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion5, @"$""""""hello + {other}""""""").WithArguments("interpolated strings", "6").WithLocation(7, 16));
+    }
+}
```

Comment:
It might be good to test permutation of different kinds of strings in different kinds of interpolation holes.
I assume you've done a search on the existing syntax kinds and syntax node types related to interpolations. Correct?

Since the binding logic wasn't affected, these suggestions are mostly exploratory testing and likely okay, but good to add or queue in test plan:
- order of evaluation (covered for interpolation handlers, but apparently not for regular interpolations) and definite assignment (`$""" {M(out var x)} {x} """` versus reverse order, for example)
- nullability: `$""" """.ToString()` is okay, passing `null` into interpolation holes
- optimization for constants
- semantic model (`GetTypeInfo`) on different parts of `var x = $""" {y} """;`, `IFormattable x = $""" """;`, `var x = $""" {null} """;`.

Reply:
> I assume you've done a search on the existing syntax kinds and syntax node types related to interpolations. Correct?

Correct. The compiler layer never checks them. 

I will add more tests for the types of interactions you mentioned.

Reply:
ok.  added tests.  there are a bunch of semantic model checks using VerifyInterpolatedStringExpression.  

could you explain this part: nullability: $""" """.ToString() is okay, passing null into interpolation holes
?

> optimization for constants

There are def test for tehse, but i can add more.

Reply:
> could you explain this part: nullability: ...

1. The result from the new expression should be not-null (so it can be dereferenced without warning).
2. Executing `$"""{null}"""`